### PR TITLE
Integrate tensorflow gradients with CppAD

### PR DESCRIPTION
**Pull request by fritzo**
**Status:** Ready to merge when reviewed. This is still buggy but failing tests have been wrapped in `expect_error`, and debugging is blocked by underlying bugs in AD.

This PR attempts to integrate Fritz's Tensorflow machinery with Nick's CppAD machinery using CppAD's [User Defined Atomic Op](https://www.coin-or.org/CppAD/Doc/atomic_base.htm) interface.

## Tasks

- [x] Add R code to add gradient ops to the Tensorflow graph before serializing it.
- [x] Expose gradient computations in the `NimTf_Runner` C++ class.
- [x] Create a wrapper class `NimTf_op` to be used in AD computations (like `NimTf_Runner`'s but for `CppAD::AD<double>` types.)
    - [x] Implement `NimTf_op::forward()` and `::reverse()` functions.
    - [x] Implement `NimTf_op::NimTf_run()` interface.
- [x] Use `NimTf_op` in the generated code for AD taping. (or merge it into `NimTf_runner`)
- [x] Get everything to compile.
- [ ] Get tests to pass.

### CODE REVIEW

File: packages/nimble/inst/CppCode/tensorflow.cpp [hunk 1]
```diff
@@ -178,3 +179,138 @@ void NimTf_Runner::NimTf_getOutput(NimArrBase<double>& nimArr) {
   TF_DeleteTensor(tensor);
   output_pos_ += 1;
 }
+
+void NimTf_Runner::NimTf_setInput(const std::vector<int>& dims,
+                                  const double* data) {
+  NIMERROR("TODO")
+}
+
+void NimTf_Runner::NimTf_runGradients() { NIMERROR("TODO") }
```
Comment (NLMichaud) [inline packages/nimble/inst/CppCode/tensorflow.cpp]: Should this be `NimTf_Runner::NimTf_runGradient()` ?

Comment (fritzo) [reply packages/nimble/inst/CppCode/tensorflow.cpp]: Good catch! (It's so nice to have someone else's eyes on the code :-)

File: packages/nimble/inst/include/nimble/tensorflow.h [hunk 2]
```diff
@@ -28,6 +28,12 @@
 #include <string>
 #include <vector>
 
+#define NIMBLE_HAVE_CPPAD 1  // DO NOT SUBMIT.
```
Comment (fritzo) [inline packages/nimble/inst/include/nimble/tensorflow.h]: Remove before submitting.

File: packages/nimble/inst/tests/test-tfize.R [hunk 3]
```diff
@@ -3,6 +3,8 @@ source(system.file(file.path('tests', 'mathTestLists.R'), package = 'nimble'))
 
 context("Testing of Tensorflow-ization")
 
+nimbleOptions(showCompilerOutput = TRUE)  # DO NOT SUBMIT
```
Comment (fritzo) [inline packages/nimble/inst/tests/test-tfize.R]: Remove before submitting.

**Comment by fritzo** [comment]
I plan to merge this Wednesday, in case anyone wants to do another review.

**Pull request merged**